### PR TITLE
Simplify Auto Bed Leveling Positions Configuration

### PR DESCRIPTION
**Pull request by tkoopman**
Calculates the ABL Grid boundaries based on bed size & probe offsets.

### CODE REVIEW

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -808,10 +808,21 @@
...
+    #define LEFT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER + PROBE_BED_MARGIN)
+    #define RIGHT_PROBE_BED_POSITION (X_MAX_POS - PROBE_BED_MARGIN)
+  #endif
+  #if X_PROBE_OFFSET_FROM_EXTRUDER < 0
+    #define FRONT_PROBE_BED_POSITION PROBE_BED_MARGIN
+    #define BACK_PROBE_BED_POSITION (Y_MAX_POS - PROBE_BED_MARGIN + Y_PROBE_OFFSET_FROM_EXTRUDER)
+  #else
+    #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER + PROBE_BED_MARGIN)
+    #define BACK_PROBE_BED_POSITION (Y_MAX_POS - PROBE_BED_MARGIN)
+  #endif
```
Comment (FHeilmann) [inline Marlin/Configuration.h]: You don't need two separaget #if #else #endif blocks if they both theck for the same thing. You can consolidate your code into one of these blocks.

Comment (tkoopman) [reply Marlin/Configuration.h]: Copy and Paste + Testing Fail. My bad :)
Tested with probe offsets both + and both -, so missed typo for when they different.
Fixed.

Comment (Blue-Marlin) [reply Marlin/Configuration.h]: I'm quite sure the second
```
#if X_PROBE_OFFSET_FROM_EXTRUDER < 0
```
should be
```
#if Y_PROBE_OFFSET_FROM_EXTRUDER < 0
```

Comment (fiveangle) [inline Marlin/Configuration.h]: I've had mine this for a while now:

```
// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
```

Comment (fiveangle) [inline Marlin/Configuration.h]: Was going to make this my first PR to get my feet wet, but you beat me to it, so please take if you think this works better.

Comment (fiveangle) [reply Marlin/Configuration.h]: I've had mine this for a while now:

```
  // Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
```

Submitting this simple PR in order to get my github feet wet has been on my todo list for quite some time and haven't gotten to it, so please take if you think this works better.

Comment (tkoopman) [reply Marlin/Configuration.h]: Looks good to me. Inline If statements neater.

Comment (los7702) [reply Marlin/Configuration.h]: Just one question? If zero is in the middle then left pos could be min probe edge to the rigt of middle...??

Comment (los7702) [reply Marlin/Configuration.h]: This is my suggestion...
```cpp
// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
#define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? X_MIN_POS + MIN_PROBE_EDGE : X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER) 
#define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
#define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? Y_MIN_POS + MIN_PROBE_EDGE : Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
#define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
```

Comment (los7702) [inline Marlin/Configuration.h]: This is my suggestion...
`// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? X_MIN_POS + MIN_PROBE_EDGE : X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? Y_MIN_POS + MIN_PROBE_EDGE : Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)`

Comment (los7702) [inline Marlin/Configuration.h]: This is my suggestion...
`// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? X_MIN_POS + MIN_PROBE_EDGE : X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? Y_MIN_POS + MIN_PROBE_EDGE : Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)`

Comment (thinkyhead) [reply Marlin/Configuration.h]: Generally all good suggestions. I think the auto-assigning of these values should be moved into `Conditionals_POST.h`, in this manner:
```cpp
#ifndef LEFT_PROBE_BED_POSITION
  #define LEFT_PROBE_BED_POSITION  ((X_PROBE_OFFSET_FROM_EXTRUDER) <= (MIN_PROBE_EDGE) ? (X_MIN_POS + MIN_PROBE_EDGE) : (X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER))
#endif
```
...and so on. (Note the extra parentheses!) Then, comment out the lines in `Configuration.h` and add documentation that these values are automatically set and only need to be changed under special conditions. Some of the example configurations might provide overrides. Most of them will probably be able to leave them undefined.

Comment (los7702) [reply Marlin/Configuration.h]: @thinkyhead that is even better. if not defined, then calculate...
1 vote from me

Comment (fiveangle) [reply Marlin/Configuration.h]: @los7702, yeah... wasn't thinking of center-origined 'bots.  Automating defaults outside of Config is perfect.

**Comment by tkoopman** [comment]
Updated based on comments. Tested fine on my printer.
How's it look to you guys now?

Tim.

**Comment by fiveangle** [comment]
Thinkyhead stated: "(Note the extra parentheses!)" but I don't see them in your code.  Probably a good idea to be doubly-sure.

**Comment by fiveangle** [comment]
Nitpicking, but you could probably change:

// The Z probe minimum outer margin (to validate G29 parameters).

to:

// The Z probe minimum outer margin (to validate G29 parameters and auto-calculate default probe bed positions).

**Comment by los7702** [comment]
Great job.
Looks good to me.

**Comment by thinkyhead** [comment]
Please only submit PRs to the `RCBugFix` branch.

**Comment by thinkyhead** [comment]
A comprehensive patch should take account of the probe offsets, and not just the `MIN_PROBE_EDGE` (which is legacy cruft). We now require C++11, so sanity checking can use `static_assert`.

**Pull request closed without merging**